flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java

```diff
@@ -54,19 +54,21 @@ default StreamStateHandle registerReference(
     }
 
     /**
-     * Register a reference to the given shared state in the registry. If there is already a state
-     * handle registered under the given key, the "new" state handle is disposed .
+     * Register a reference to the given shared state in the registry, the registry key should be
+     * based on the physical identification of the state. If there is already a state handle
+     * registered under the given key and the 'new' state is not equal to the old one, an exception
+     * will be thrown.
      *
-     * <p>IMPORTANT: caller should check the state handle returned by the result, because the
-     * registry is performing de-duplication and could potentially return a handle that is supposed
-     * to replace the one from the registration request.
+     * <p>IMPORTANT: After FLINK-29913, caller only needs to check the state handle returned by the
+     * result only if the 'new' state is a placeholder, because the registry is no longer performing
+     * de-duplication.
```

Comment:
I think this is an implementation detail about the handle types.
I'd rather suggest to always check the result, so that we don't need to update callers if a new replacement is introduced:
```suggestion
     * <p>IMPORTANT: the caller must use the returned state handle instead of the passed one
     * because the registry might replace or update it.
```